### ISSUE TITLE
how to make custom eval dataset  

### ISSUE THREAD

**Issue by jakhon37** [mk-minchul/CVLface#17]
Hi, I'm working on creating a custom training and evaluation dataset for animation faces. Should the class indices in the evaluation dataset exactly match those in the training dataset? Also, could you provide some tips for building an effective custom evaluation dataset?

**Comment by OlivierLienhard** [comment]
Hello Jakhon,

I asked me the sames questions, and i believe the answer to the first question is no, based on how i understood the way the evaluation is done.
I tried a way to create a custom evaluation dataset, based on the evaluation dataset given in this repo, and I ended up with the following code :

```
def create_verification_csv(root_dir, output_csv, num_pairs_per_person=5, negative_ratio=1):

    all_images = []
    person_to_images = {}

    # Collect all images and organize them by person
    for person_folder in os.listdir(root_dir):
        person_path = os.path.join(root_dir, person_folder)
        if os.path.isdir(person_path):
            images = [
                os.path.join(person_folder, img)
                for img in os.listdir(person_path)
                if img.lower().endswith((".png", ".jpg", ".jpeg"))
            ]
            all_images.extend(images)
            person_to_images[person_folder] = images

    pairs = []
    # Create positive pairs (same person)
    for _, images in person_to_images.items():
        if len(images) < 2:
            continue
        pairs.extend(
            random.sample(
                list(combinations(images, 2)),
                min(num_pairs_per_person, len(list(combinations(images, 2)))),
            )
        )

    # Create negative pairs (different people)
    num_negative_pairs = int(len(pairs) * negative_ratio)
    negative_pairs = set()
    while len(negative_pairs) < num_negative_pairs:
        img1, img2 = random.sample(all_images, 2)
        if os.path.dirname(img1) != os.path.dirname(img2):
            negative_pairs.add((img1, img2))

    # Write to CSV
    with open(output_csv, "w", newline="") as csvfile:
        csvwriter = csv.writer(csvfile)
        csvwriter.writerow(["image1", "image2", "is_same"])
        # Combine positive and negative pairs
        all_pairs = list(zip(pairs, negative_pairs))
        random.shuffle(all_pairs)

        for pos_pair, neg_pair in all_pairs:
            # Write positive pair
            csvwriter.writerow([pos_pair[0], pos_pair[1], True])
            # Write negative pair
            csvwriter.writerow([neg_pair[0], neg_pair[1], False])

    print(f"CSV file created at {output_csv}")
    print(f"Total pairs: {len(pairs) + len(negative_pairs)}")
    print(f"Positive pairs: {len(pairs)}")
    print(f"Negative pairs: {len(negative_pairs)}")

def create_verification_dataset(csv_path, image_dir):
    # Read the CSV file
    df = pd.read_csv(csv_path)

    # Create lists to hold our data
    image_paths = []
    is_same_list = []
    index_list = []

    # Populate the lists
    for idx, row in df.iterrows():
        image_paths.append(os.path.join(image_dir, row["image1"]))
        image_paths.append(os.path.join(image_dir, row["image2"]))
        is_same_list.append(row["is_same"])
        is_same_list.append(row["is_same"])
        index_list.append(idx * 2)  # Use the DataFrame index as our pair index
        index_list.append(idx * 2 + 1)  # Use the DataFrame index as our pair index

    # Create the dataset
    dataset = Dataset.from_dict(
        {
            "image": image_paths,
            "is_same": is_same_list,
            "index": index_list,
        }
    )

    # Load images
    dataset = dataset.cast_column("image", Image())

    return dataset


# Usage
root_dir = "/home/olivier/data/validation_folders"
output_csv = "verification_pairs.csv"
create_verification_csv(
    root_dir, output_csv, num_pairs_per_person=5, negative_ratio=1
)
dataset = create_verification_dataset("verification_pairs.csv", root_dir)
print(dataset)
dataset.save_to_disk(
    "/home/olivier/data/facerec_val/custom_val"
)

```

In this code, the root_dir at the end is the custom repo where you have a a hierarchy of folders like this,  each folder containing all images of the same person (or class) : 

- validation_folders/person1/ 
- validation_folders/person2/
- validation_folders/person3/
- ...

The first function then creates a csv containing positive and negative pairs, and the second simply converts the csv to the same format as needed for this repo, which we save to disk.
Finally, you have to modify slightly the yaml in the evaluations folder you call when you run your scripts, with something like this : 
```
per_epoch_evaluations: {
  "custom_val": {
    'path': 'facerec_val/custom_val',
    'evaluation_type': 'verification',
    'color_space': 'RGB',
    'batch_size': 32,
    'num_workers': 4
   }
}
```

I just tested it and my results seem OK on my use case, so I put it here to help but I would be happy if anyone can spot any error in this or has any other way to do it.

**Comment by OlivierLienhard** [comment]
Just saw there was already a closed issue in this #5  . Maybe this is what you were looking for.

**Comment by jakhon37** [comment]
Hi @OlivierLienhard. Thank you for your help with appreciation.
